### PR TITLE
[QA-1895] Integration test: speculative fix for navigation timeout error

### PR DESCRIPTION
**Pull request by aweng98**
https://broadworkbench.atlassian.net/browse/QA-1895

Seeing sporadic UI test failures caused by Navigation timeout. 

This was recently added
```
const waitUntilLoadedOrTimeout = timeout => ({ waitUntil: ['load', 'domcontentloaded', 'networkidle0'], timeout })
```

I think it has to do with `waitUntil: load, domcontentloaded` options so I'm going to remove these two.

### CODE REVIEW

File: integration-tests/utils/integration-utils.js [hunk 1]
```diff
@@ -346,7 +353,7 @@ const withPageLogging = fn => async options => {
   return await fn(options)
 }
 
-const waitUntilLoadedOrTimeout = timeout => ({ waitUntil: ['load', 'domcontentloaded', 'networkidle0'], timeout })
+const waitUntilNetworkIdleOrTimeout = timeout => ({ waitUntil: ['networkidle0'], timeout })
```
Comment (aweng98) [inline integration-tests/utils/integration-utils.js:356]: I'm renaming function accordingly since removing `load` and `domcontentloaded` options,

File: integration-tests/utils/integration-utils.js [hunk 2]
```diff
@@ -159,8 +159,13 @@ const select = async (page, labelContains, text) => {
   return click(page, `//div[starts-with(@id, "react-select-") and contains(normalize-space(.),"${text}")]`)
 }
 
-const waitForNoSpinners = page => {
-  return page.waitForXPath('//*[@data-icon="loadingSpinner"]', { hidden: true })
+const waitUntilDomContentLoaded = page => {
+  return page.waitForSelector('div:not(empty):not([id="root"])') // Used to prevent checking loadingSpinner while page is blank
```
Comment (aweng98) [inline integration-tests/utils/integration-utils.js]: This CSS selector should be a good start for most pages but probably not sufficient for edge cases.

File: integration-tests/utils/integration-utils.js [hunk 3]
```diff
@@ -159,8 +159,14 @@ const select = async (page, labelContains, text) => {
   return click(page, `//div[starts-with(@id, "react-select-") and contains(normalize-space(.),"${text}")]`)
 }
 
-const waitForNoSpinners = page => {
-  return page.waitForXPath('//*[@data-icon="loadingSpinner"]', { hidden: true })
+// Used to prevent checking loadingSpinner while page is blank
+const waitUntilDomContentLoaded = page => {
+  return page.waitForSelector('div:not(empty):not([id="root"])')
```
Comment (aweng98) [inline integration-tests/utils/integration-utils.js]: This CSS selector should be a good start for most pages but probably not sufficient for edge cases.

Comment (nawatts) [reply integration-tests/utils/integration-utils.js]: If I understand this correctly, the purpose of this is to wait for the initial "Loading Terra..." message to go away? If so, the name seems misleading. Based on the name, I would have assumed this had something to do with the [DOMContentLoaded event](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event).

Also, this selector looks like it matches the loading message itself, so wouldn't this function return while the loading message is still visible?
https://github.com/DataBiosphere/terra-ui/blob/2c58e0f146245c119c5cdf4d1732591bad950ab1/public/index.html#L17-L30

Finally, was the selector meant to be `:not(:empty)` vs `:not(empty)`?

File: integration-tests/utils/integration-utils.js [hunk 4]
```diff
@@ -192,6 +198,7 @@ const dismissNotifications = async page => {
 }
 
 const signIntoTerra = async (page, { token, testUrl }) => {
+  console.log('signIntoTerra ...')
```
Comment (nawatts) [inline integration-tests/utils/integration-utils.js:195]: Is this intentional or left over from development?

File: integration-tests/utils/integration-utils.js (hunk 1 above)
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: I realize the old function was also called `waitUntil...`, but this name seems misleading. Based on the name, I would have assumed that this waits for the network to be idle. But it actually just returns options to be passed to another function which does the waiting.

**Comment by nawatts** [comment]
Not necessarily for this PR, but based on the number of `page.goto` calls that had to be changed here, maybe a `goToPage` helper function would nice. So that timeout options don't have to be specified every time.

File: integration-tests/utils/integration-utils.js (hunk 3 above)
Comment (aweng98) [reply integration-tests/utils/integration-utils.js]: Purpose of `waitUntilDomContentLoaded` function is wait until page is not blank otherwise time out, to prevent checking for `Loading Terra ...` spinner too early.
It doesn't matter if the selector is finding the loading message so yes, this function returns while loading message is still visible. Waiting for spinner to stop is done by `await page.waitForXPath('//*[@data-icon="loadingSpinner"]', { hidden: true })`. 
See `waitForNoSpinners` function (below).

File: integration-tests/utils/integration-utils.js (hunk 4 above)
Comment (aweng98) [reply integration-tests/utils/integration-utils.js:195]: Intentional. It makes easier (hopefully for everyone) to follow test events when reading failed test log.

File: integration-tests/utils/integration-utils.js (hunk 1 above)
Comment (aweng98) [reply integration-tests/utils/integration-utils.js]: Good point. What do you think about `navigationOptionNetworkIdle` as new name?

**Review by aweng98 — COMMENTED** [review]
👍🏻 `:not(:empty)` looks to me is a better selector. On home page `https://bvdp-saturn-staging.appspot.com/`, `:not(empty)` found 70 web elements. `:not(:empty)` found 51 web elements. 70 web elements included empty div element like `<div></div>` which is "bad".

**Comment by aweng98** [comment]
Can you think a better way to hold off checking loading spinner until page is "ready" besides checking for blank page?

File: integration-tests/utils/integration-utils.js [hunk 5]
```diff
@@ -159,8 +159,17 @@ const select = async (page, labelContains, text) => {
   return click(page, `//div[starts-with(@id, "react-select-") and contains(normalize-space(.),"${text}")]`)
 }
 
-const waitForNoSpinners = page => {
-  return page.waitForXPath('//*[@data-icon="loadingSpinner"]', { hidden: true })
+// Used to prevent checking loadingSpinner while page is blank
+const waitUntilDomContentLoaded = page => {
+  return Promise.all([
+    page.waitForFunction(() => document),
+    page.waitForSelector('div:not(:empty):not([id="root"])')
```
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: I still don't understand this function. This selector matches the initial "Loading Terra..." message as well as anything rendered after that message goes away. So why is this complex selector needed? It seems like we could just wait for the root div with `#root"`.

File: integration-tests/utils/integration-utils.js [hunk 6]
```diff
@@ -159,8 +159,17 @@ const select = async (page, labelContains, text) => {
...
+  ])
+}
+
+const waitForNoSpinners = async page => {
+  await waitUntilDomContentLoaded(page)
```
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: It seems like `waitUntilDomContentLoaded` would only be needed on the initial page load. Since `waitForNoSpinners` is frequently used, is it worth calling `waitUntilDomContentLoaded` every time? Would putting it in `goToPage` work?

File: integration-tests/utils/integration-utils.js (hunk 5 above)
Comment (aweng98) [reply integration-tests/utils/integration-utils.js]: I have removed `waitUntilDomContentLoaded`.

File: integration-tests/utils/integration-helpers.js [hunk 7]
```diff
@@ -34,6 +33,8 @@ const makeWorkspace = withSignedInPage(async ({ page, billingProject }) => {
       try {
         return await window.Ajax().Workspaces.create({ namespace: billingProject, name, attributes: {} })
       } catch (err) {
+        console.error(err)
+        console.error(typeof err)
```
Comment (aweng98) [inline integration-tests/utils/integration-helpers.js:37]: I'd like to use this log to determine why `await err.text()` fails sometimes.

Comment (nawatts) [reply integration-tests/utils/integration-helpers.js:37]: Isn't this already logged below? With:

```js
console.error(`Failed to create workspace: ${workspaceName} with billing project: ${billingProject}`)
console.error(e)
```

File: integration-tests/utils/integration-utils.js [hunk 8]
```diff
@@ -346,7 +347,11 @@ const withPageLogging = fn => async options => {
   return await fn(options)
 }
 
-const waitUntilLoadedOrTimeout = timeout => ({ waitUntil: ['load', 'domcontentloaded', 'networkidle0'], timeout })
+const navOptionNetworkIdle = (timeout = 60 * 1000) => ({ waitUntil: ['domcontentloaded'], timeout })
```
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: The PR description says:
> Seeing sporadic UI test failures caused by Navigation timeout.
> ...
> I think it has to do with waitUntil: load, domcontentloaded options so I'm going to remove these two.

But this removes `load` and `networkidle0`, not `domcontentloaded`. What changed?

Also, the function name is misleading since `navOptionNetworkIdle` does not wait for `networkidle`.

Comment (aweng98) [reply integration-tests/utils/integration-utils.js]: Oops! I think I just forgot to change it back. it should be `networkidle0`.

File: integration-tests/utils/integration-helpers.js (hunk 7 above)
Comment (aweng98) [reply integration-tests/utils/integration-helpers.js:37]: No. because that catches error caused by `await err.text()`. I want to see what the original `err` is and what the `typeof` is.

**Review by nawatts — APPROVED** [review]

**Review by cahrens — APPROVED** [review]
